packages/babel-plugin-class-features/src/fields.js

```diff
@@ -0,0 +1,315 @@
+      template.expression`BASE(REF, PROP)[PROP]`({
+        BASE: file.addHelper("classPrivateFieldLooseBase"),
+        REF: object,
+        PROP: privateNamesMap.get(name).id,
+      }),
+    );
+  },
+};
+
+export function transformPrivateNamesUsage(
+  ref,
+  path,
+  privateNamesMap,
+  loose,
+  state,
+) {
+  const body = path.get("body");
+
+  if (loose) {
+    body.traverse(privateNameVisitor, {
+      privateNamesMap,
+      file: state,
+      ...privateNameHandlerLoose,
+    });
+  } else {
+    memberExpressionToFunctions(body, privateNameVisitor, {
+      privateNamesMap,
+      classRef: ref,
+      file: state,
+      ...privateNameHandlerSpec,
+    });
+  }
+}
+
+function buildPrivateFieldInitLoose(ref, prop, privateNamesMap) {
+  const { id } = privateNamesMap.get(prop.node.key.id.name);
+  const value = prop.node.value || prop.scope.buildUndefinedNode();
+
+  return template.statement.ast`
+    Object.defineProperty(${ref}, ${id}, {
+      // configurable is false by default
+      // enumerable is false by default
+      writable: true,
+      value: ${value}
+    });
+  `;
+}
+
+function buildPrivateInstanceFieldInitSpec(ref, prop, privateNamesMap) {
+  const { id } = privateNamesMap.get(prop.node.key.id.name);
+  const value = prop.node.value || prop.scope.buildUndefinedNode();
+
+  return template.statement.ast`${id}.set(${ref}, {
+    // configurable is always false for private elements
+    // enumerable is always false for private elements
+    writable: true,
+    value: ${value},
+  })`;
+}
+
+function buildPrivateStaticFieldInitSpec(prop, privateNamesMap) {
+  const { id } = privateNamesMap.get(prop.node.key.id.name);
+  const value = prop.node.value || prop.scope.buildUndefinedNode();
+
+  return template.statement.ast`
+    var ${id} = {
+      // configurable is false by default
+      // enumerable is false by default
+      writable: true,
+      value: ${value}
+    };
+  `;
+}
+
+function buildPublicFieldInitLoose(ref, prop) {
+  const { key, computed } = prop.node;
+  const value = prop.node.value || prop.scope.buildUndefinedNode();
+
+  return t.expressionStatement(
+    t.assignmentExpression(
+      "=",
+      t.memberExpression(ref, key, computed || t.isLiteral(key)),
+      value,
+    ),
+  );
+}
+
+function buildPublicFieldInitSpec(ref, prop, state) {
+  const { key, computed } = prop.node;
+  const value = prop.node.value || prop.scope.buildUndefinedNode();
+
+  return t.expressionStatement(
+    t.callExpression(state.addHelper("defineProperty"), [
+      ref,
+      computed || t.isLiteral(key) ? key : t.stringLiteral(key.name),
+      value,
+    ]),
+  );
+}
+
+export function buildFieldsInitNodes(
+  ref,
+  props,
+  privateNamesMap,
+  state,
+  loose,
+) {
+  const staticNodes = [];
+  const instanceNodes = [];
+
+  for (const prop of props) {
+    const isStatic = prop.node.static;
+    const isPrivate = prop.isPrivate();
+
+    // Pattern matching please
+    switch (true) {
```

Comment:
lol, TIL

Reply:
is this really more readable than cascading if/else chain though?

Reply:
I find `switch` more readable than long if/else chains with small bodies, but I don't have anything against using if/else here.

<details>
<summary>switch(true)</summary>

```js
    switch (true) {
      case isStatic && isPrivate && loose:
        staticNodes.push(
          buildPrivateFieldInitLoose(t.cloneNode(ref), prop, privateNamesMap),
        );
        break;
      case isStatic && isPrivate && !loose:
        staticNodes.push(
          buildPrivateStaticFieldInitSpec(prop, privateNamesMap),
        );
        break;
      case isStatic && !isPrivate && loose:
        staticNodes.push(buildPublicFieldInitLoose(t.cloneNode(ref), prop));
        break;
      case isStatic && !isPrivate && !loose:
        staticNodes.push(
          buildPublicFieldInitSpec(t.cloneNode(ref), prop, state),
        );
        break;
      case !isStatic && isPrivate && loose:
        instanceNodes.push(
          buildPrivateFieldInitLoose(t.thisExpression(), prop, privateNamesMap),
        );
        break;
      case !isStatic && isPrivate && !loose:
        instanceNodes.push(
          buildPrivateInstanceFieldInitSpec(
            t.thisExpression(),
            prop,
            privateNamesMap,
          ),
        );
        break;
      case !isStatic && !isPrivate && loose:
        instanceNodes.push(buildPublicFieldInitLoose(t.thisExpression(), prop));
        break;
      case !isStatic && !isPrivate && !loose:
        instanceNodes.push(
          buildPublicFieldInitSpec(t.thisExpression(), prop, state),
        );
        break;
      default:
        throw new Error("Unreachable.");
    }
```

</details>

<details>
<summary>If/else</summary>

```js
    if (isStatic && isPrivate && loose) {
      staticNodes.push(
        buildPrivateFieldInitLoose(t.cloneNode(ref), prop, privateNamesMap),
      );
    } else if (isStatic && isPrivate && !loose) {
      staticNodes.push(buildPrivateStaticFieldInitSpec(prop, privateNamesMap));
    } else if (isStatic && !isPrivate && loose) {
      staticNodes.push(buildPublicFieldInitLoose(t.cloneNode(ref), prop));
    } else if (isStatic && !isPrivate && !loose) {
      staticNodes.push(buildPublicFieldInitSpec(t.cloneNode(ref), prop, state));
    } else if (!isStatic && isPrivate && loose) {
      instanceNodes.push(
        buildPrivateFieldInitLoose(t.thisExpression(), prop, privateNamesMap),
      );
    } else if (!isStatic && isPrivate && !loose) {
      instanceNodes.push(
        buildPrivateInstanceFieldInitSpec(
          t.thisExpression(),
          prop,
          privateNamesMap,
        ),
      );
    } else if (!isStatic && !isPrivate && loose) {
      instanceNodes.push(buildPublicFieldInitLoose(t.thisExpression(), prop));
    } else if (!isStatic && !isPrivate && !loose) {
      instanceNodes.push(
        buildPublicFieldInitSpec(t.thisExpression(), prop, state),
      );
    } else {
      throw new Error("Unreachable.");
    }
```

</details>

Reply:
i guess the switch version is more readable indeed, putting aside the weird (to me) shape of this switch/case